### PR TITLE
Implement KeOps Integration

### PR DESCRIPTION
**Pull request by jacobrgardner**
This PR implements two new kernels `kernels.keops.RBFKernel` and `kernels.keops.MaternKernel`, that when swapped in for our default ones use KeOps to drive kernel matmuls.

Initial results suggest that the efficient matmuls of KeOps combined with the GPyTorch bells and whistles like preconditioning allow for training on datasets like 3droad (`n=~400k`) in minutes.

### CODE REVIEW

**Comment by gpleiss** [comment]
Cc/ @jeanfeydy @joanglaunes @bcharlier

**Comment by jacobrgardner** [comment]
Once this goes in, I'll work on adding a few more example notebooks to the folder and slowly build out the integration that way. In particular, here are three more notebooks I'd like to add (probably in the order that I'll tackle them):

1. Integrating multi GPU support.
2. Integrating KeOps in to a sparse or variational GP model.
3. Integrationg KeOps in to a Deep GP model.

**Comment by jacobrgardner** [comment]
@gpleiss No idea why the docs are failing -- they are failing on master too, and they do build locally without any problems.

File: gpytorch/kernels/keops/matern_kernel.py [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+                if self.nu == 0.5:
+                    constant_component = 1
+                elif self.nu == 1.5:
+                    constant_component = (math.sqrt(3) * distance).add(1)
+                elif self.nu == 2.5:
+                    constant_component = (math.sqrt(5) * distance).add(1).add(5.0 / 3.0 * distance ** 2)
+                return constant_component * exp_component
+            else:
+                with torch.autograd.enable_grad():
```
Comment (gpleiss) [inline gpytorch/kernels/keops/matern_kernel.py]: Why is this necessary?

File: gpytorch/lazy/keops_lazy_tensor.py [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+        return self.covar_func(self.x1, self.x2, **self.params)
+
+    def _matmul(self, rhs):
+        # TODO: replace with `self.covar_mat @ rhs` on next PyKeOps release.
+        batch_shape = self.batch_shape
+        if len(batch_shape):
+            # Equivalent to dim=-3, but KeOps LT doesn't support relative index in sum.
+            return (self.covar_mat * rhs[..., None, :, :].contiguous()).sum(dim=len(batch_shape) + 1)
+        else:
+            if not rhs.is_contiguous():
```
Comment (gpleiss) [inline gpytorch/lazy/keops_lazy_tensor.py]: Nit: this is not necessary - `mat.contiguous` returns `mat` if `mat.is_contiguous`.
